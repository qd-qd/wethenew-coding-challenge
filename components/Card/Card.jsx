import Link from 'next/link';
import PropTypes from 'prop-types';

import styles from './Card.module.css';

const Card = ({ slug, image, price, name, brand }) => (
  <Link href="/[slug]" as={`/${slug}`}>
    <a data-testid="card-link">
      <div className={styles.container}>
        <img
          src={image}
          alt={`The ${name} sneaker by ${brand}`}
          className={styles.image}
        />
        <div className={styles['section-infos']}>
          <p className={styles.name}>{name}</p>
          <p className={styles.price}>
            {price}
            <span>â‚¬</span>
          </p>
        </div>
        <p className={styles.brand}>{brand}</p>
      </div>
    </a>
  </Link>
);

Card.defaultProps = {
  image: '/sneakers-placeholder.png',
};

Card.propTypes = {
  slug: PropTypes.string.isRequired,
  price: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  brand: PropTypes.string.isRequired,
  image: PropTypes.string,
};

export default Card;
