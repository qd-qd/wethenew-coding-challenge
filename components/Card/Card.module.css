.container {
  display: flex;
  flex-direction: column;
  width: 350px;
  height: 400px;
  background-color: var(--theme-color-white);

  box-shadow: var(--theme-elevation-1);
  border-radius: var(--theme-shape-radius);
  overflow: hidden;
  color: var(--theme-color-text-primary);
}

.container:hover {
  box-shadow: var(--theme-elevation-4);
}

.image {
  object-fit: cover;
  flex-grow: 5;
  width: 100%;
}

.section-infos {
  display: flex;
  justify-content: space-between;
  align-items: baseline;
  padding: 0.5rem;
  border-top: 1px solid var(--theme-color-grey-200);
}

.name {
  width: 25ch;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  font-size: 1.1rem;
  font-family: var(--theme-font-family);
  font-weight: var(--theme-font-weight-light);
  text-decoration: none;
}

.section-infos:hover .name {
  width: auto;
  text-overflow: unset;
  white-space: unset;
  overflow: unset;
}

.price {
  font-size: 1.2rem;
  font-family: var(--theme-font-family);
  font-weight: var(--theme-font-weight-regular);
  text-decoration: none;
}

.brand {
  position: absolute;
  padding: 0.25rem 0.5rem;
  background-color: var(--theme-color-grey-200);
  color: var(--theme-color-grey-800);
  margin: 0.5rem 0 0 0.5rem;
  border-radius: 12px;
}
