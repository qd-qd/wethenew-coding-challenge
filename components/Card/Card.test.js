import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import sneaker from '@test/fixtures/sneaker.json';
import Card from './Card';

describe('Card', () => {
  test('displays greeting', () => {
    render(
      <Card
        slug={sneaker.slug}
        image={sneaker.image}
        price={sneaker.price}
        name={sneaker.name}
        brand={sneaker.brand}
      />,
    );

    expect(screen.getByTestId('card-link')).toBeDefined();
    expect(screen.getByTestId('card-link').href).toContain(sneaker.slug);
    expect(
      screen.getByAltText(`The ${sneaker.name} sneaker by ${sneaker.brand}`),
    ).toBeTruthy();
    expect(screen.getByRole('img').src).toContain(sneaker.image);
    expect(screen.getByRole('img').className).toBeTruthy();
    expect(screen.getByRole('img').alt).toBeTruthy();

    expect(screen.getByText(sneaker.name)).toHaveAttribute('class');
    expect(screen.getByText(sneaker.price)).toHaveAttribute('class');
    expect(screen.getByText(sneaker.brand)).toHaveAttribute('class');
    expect(screen.getByText('â‚¬')).toBeTruthy();
  });

  test('displays placeholder image if no one is pass', () => {
    render(
      <Card
        slug={sneaker.slug}
        price={sneaker.price}
        name={sneaker.name}
        brand={sneaker.brand}
      />,
    );
    expect(screen.getByRole('img').src).toContain('placeholder');
  });

  test("doesn't break snapshot", () => {
    const { container } = render(
      <Card
        slug={sneaker.slug}
        image={sneaker.image}
        price={sneaker.price}
        name={sneaker.name}
        brand={sneaker.brand}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
