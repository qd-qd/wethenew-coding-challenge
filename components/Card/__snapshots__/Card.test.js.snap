// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card doesn't break snapshot 1`] = `
<div>
  <a
    data-testid="card-link"
    href="/fake-url"
  >
    <div
      class="container"
    >
      <img
        alt="The Fake shoes sneaker by Jest dude"
        class="image"
        src="www.fakeimage.com/toto.jpg"
      />
      <div
        class="section-infos"
      >
        <p
          class="name"
        >
          Fake shoes
        </p>
        <p
          class="price"
        >
          999
          <span>
            â‚¬
          </span>
        </p>
      </div>
      <p
        class="brand"
      >
        Jest dude
      </p>
    </div>
  </a>
</div>
`;
